libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -602,64 +600,44 @@ bool TypeInference::visit(TypeClassInstantiation const& _typeClassInstantiation)
 	auto& instantiationAnnotation = annotation(_typeClassInstantiation);
 	if (instantiationAnnotation.type)
 		return false;
+
 	instantiationAnnotation.type = m_voidType;
-	std::optional<TypeClass> typeClass = std::visit(util::GenericVisitor{
-		[&](ASTPointer<IdentifierPath> _typeClassName) -> std::optional<TypeClass> {
-			if (auto const* typeClassDefinition = dynamic_cast<TypeClassDefinition const*>(_typeClassName->annotation().referencedDeclaration))
-			{
-				// visiting the type class will re-visit this instantiation
-				typeClassDefinition->accept(*this);
-				// TODO: more error handling? Should be covered by the visit above.
-				solAssert(m_analysis.annotation<TypeClassRegistration>(*typeClassDefinition).typeClass.has_value());
-				return m_analysis.annotation<TypeClassRegistration>(*typeClassDefinition).typeClass.value();
-			}
-			else
-			{
-				m_errorReporter.typeError(9817_error, _typeClassInstantiation.typeClass().location(), "Expected type class.");
-				return std::nullopt;
-			}
-		},
-		[&](Token _token) -> std::optional<TypeClass> {
-			auto const& classRegistrationAnnotation = m_analysis.annotation<TypeClassRegistration>();
-			if (auto builtinClass = builtinClassFromToken(_token))
-				if (auto typeClass = util::valueOrNullptr(classRegistrationAnnotation.builtinClasses, *builtinClass))
-					return *typeClass;
-			m_errorReporter.typeError(2658_error, _typeClassInstantiation.location(), "Invalid type class name.");
-			return std::nullopt;
-		}
-	}, _typeClassInstantiation.typeClass().name());
-	if (!typeClass)
-		return false;
 
-	// TODO: _typeClassInstantiation.typeConstructor().accept(*this); ?
-	auto typeConstructor = m_analysis.annotation<TypeRegistration>(_typeClassInstantiation.typeConstructor()).typeConstructor;
-	if (!typeConstructor)
-	{
-		m_errorReporter.typeError(2138_error, _typeClassInstantiation.typeConstructor().location(), "Invalid type constructor.");
-		return false;
-	}
+	auto const& annotationFromInstantiationRegistration = m_analysis.annotation<InstantiationRegistration>(_typeClassInstantiation);
+	auto const& nameAnnotationFromRegistration = m_analysis.annotation<InstantiationRegistration>(_typeClassInstantiation.typeClass());
 
-	std::vector<Type> arguments;
-	Arity arity{
-		{},
-		*typeClass
-	};
+	solAssert(annotationFromInstantiationRegistration.instanceType.has_value());
+	Type instanceType = *annotationFromInstantiationRegistration.instanceType;
 
+	TypeClassDefinition const* typeClassDefinition = nameAnnotationFromRegistration.typeClassDefinition;
+	if (typeClassDefinition)
+		// We may have not visited the type class definition yet. Do it just in case.
+		// TODO: Type classes should be processed in a separate pass to avoid this out-of-order visitation.
+		typeClassDefinition->accept(*this);
```

Comment:
Note that we still need to visit type class *definitions* out of order to get members registered. This will become unnecessary once we have a separate class member registration pass. #14578 was an early attempt to do that but I'm going to redo it on top of #14721 eventually.